### ISSUE TITLE
AMD getting GL_NV_MESH_SHADER Equivalent Next quarter.

### ISSUE THREAD

**Issue by IGN-Styly** [MCRcortex/nvidium#191]
[OpenGL issue](https://github.com/KhronosGroup/OpenGL-API/issues/93)
AMD graphics cards should soon be compatible with this mod, this mod may require minor changes for the AMD implementation. I'd like to hear the Devs opinion on this.

**Comment by ImBehinYa** [comment]
I opened an issue in AMD's new driver forum on GitHub and already received a response. They are aiming for the end of the year. Maybe you should comment or react so they see that this matters to us.

https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4

**Comment by Eiion** [comment]
An introduction video to Mesh Shaders directly from AMD with the RNDA 3 architecture:
https://www.youtube.com/watch?v=MQv76-q2cm8

**Comment by IGN-Styly** [comment]
Mesh shaders should be supported RDNA 2=<, after digging on my gpu firmware there seems to some mesh shader process.

**Comment by Eiion** [comment]
From all I know it - at this point - is currently not supported. Sure there may be fractures in the firmware, which is no surprise considering they are working on it, but it is not supported and code should not work. Yet.

**Comment by IGN-Styly** [comment]
> .

Firmware comes default, i havent updated it ever. The only missing piece is the driver and opengl extension, hardware wise its good to go.

**Comment by Eiion** [comment]
So you've bought a GPU the firmware off which already has fractures of code on it. Just like most firmware versions partially has code of future firmware versions left in it. That's all that is saying.
The point is, mesh shaders are currently not working on AMD cards. Yet. Nobody will put the effort into creating/completing firmware and drivers to make it work if AMD won't do it. That's not even an argument to make.

**Comment by Weather-OS** [comment]
> So you've bought a GPU the firmware off which already has fractures of code on it. Just like most firmware versions partially has code of future firmware versions left in it. That's all that is saying. The point is, mesh shaders are currently not working on AMD cards. Yet. Nobody will put the effort into creating/completing firmware and drivers to make it work if AMD won't do it. That's not even an argument to make.

Umm, you may be interested in [what GPUs support VK_EXT_mesh_shader](https://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_EXT_mesh_shader&platform=all).

**Comment by IGN-Styly** [comment]
According to the provided link 6000 series do support Mesh shader, ie can be tested easily with vulcan.

**Comment by BestOfAfrica** [comment]
> According to the provided link 6000 series do support Mesh shader, ie can be tested easily with vulcan.

Open GL.

**Comment by IGN-Styly** [comment]
> .

same pipeline.

**Comment by Quantum39** [comment]
opengl and vulkan are not the same at all...

**Comment by IGN-Styly** [comment]
They use the same gpu pipeline though different they are using the same silicon.

**Comment by Weather-OS** [comment]
> They use the same gpu pipeline though different they are using the same silicon.

Define what a “pipeline” is.

**Comment by Weather-OS** [comment]
You can’t just use Vulkan extensions inside OpenGL applications without some sort of interoperability layer. Even then, it would be excruciatingly difficult to do so.

**Comment by Quantum39** [comment]
even if they are they same pipeline, vulkan and opengl are two entirely different apis, and as such and not compatible with each other. You simply cannot run vulkan extensions inside an opengl program, or opengl extensions inside vulkan, like @Weather-OS said, without some layer or just completely reprogramming it.

**Comment by IGN-Styly** [comment]
what is a pipeline?
the pipeline is what Opengl or vulkan uses in the GPU to process instructions.
It dosen't matter what api its using as long as the silicon and firmware support it.

**Comment by CrabbyDisk** [comment]
the thing that you need to support OpenGL or Vulkan, and doing  that is hard. If you want to support OpenGL, you have to do all that work even if you have already done it for Vulkan.

**Comment by Quantum39** [comment]
the firmware and silicon supports it, however the api does not, so it will not work. Api must support the feature for it to be used, else you have 2 choices; switch apis and reprogram, or wait until x feature is supported in the api.

**Comment by Weather-OS** [comment]
> what is a pipeline?
> the pipeline is what Opengl or vulkan uses in the GPU to process instructions.
> It dosen't matter what api its using as long as the silicon and firmware support it.

That is wildly misleading.
A graphics API is made to abstract an interface between software and user-space graphics drivers. What you’re basically saying is that we should completely ignore all of that and call internal functions from user-space libraries by hand and somehow create a working solution. This is like resorting to kernel functions for printing “Hello, World!” on a terminal. Not to mention nearly impossible to do because there are no header files so you’d need to use a foreign function interface.

**Comment by IGN-Styly** [comment]
> It dosen't matter what api its using as long as the silicon and firmware support it.

it is quite an oversimplification but the point is that what Vulkan is using for mesh shaders is likely the same or similar being used by OpenGL.

**Comment by IGN-Styly** [comment]
Good Morning @MCRcortex ,
 could you lock the the issue? until anything relevant comes.

**Comment by Philanche** [comment]
> > It dosen't matter what api its using as long as the silicon and firmware support it.
> 
> it is quite an oversimplification but the point is that what Vulkan is using for mesh shaders is likely the same or similar being used by OpenGL.

It's not an oversimplification, it's simply wrong. Yes, the way that the GPU's hardware does things will be the same, but in order to make the GPU do it you need to go through some interface, which is what OpenGL and Vulkan are. But you can't mix and match, you need to use only one of them, and in this case Minecraft uses OpenGL so nvidium needs to go through that interface, but there currently is no way to do it for AMD. Once they'll have implemented it, then yeah probably. 
In general, if you don't know what you're talking about, please don't just keep making things up. (And you clearly don't, otherwise you wouldn't be saying "likely")

**Comment by Samalando** [comment]
This is like saying a mod made for fabric automatically works on forge because they both use Minecraft

**Comment by IGN-Styly** [comment]
@Samalando, thats a different topic and its not impossible(its a question of effort).

**Comment by IGN-Styly** [comment]
@Philanche i agree with you, i did make a mistake trying to explain it in simple terms. But correct me if I am wrong.
There's the GPU there the API you use to communicate with it(Vulkan or Opengl) which both use the same silicon.

**Comment by CrabbyDisk** [comment]
> @Samalando, thats a different topic and its not impossible(its a question of effort).

You just said it, it's a question of effort. Would you have the time and skill to rewrite the entire minecraft renderer with a completely different graphics API just to add mesh shader support to AMD gpus.

**Comment by Weather-OS** [comment]
> @Philanche i agree with you, i did make a mistake trying to explain it in simple terms. But correct me if I am wrong.
> There's the GPU there the API you use to communicate with it(Vulkan or Opengl) which both use the same silicon.

Your problem here is that you’re abstracting way too much detail here. It’s like me saying “there are CPU architectures like x86 or ARM which they both use electricity”.
Graphics APIs don’t directly communicate with the die of a GPU. They go through many abstraction layers to reach that point and that’s the thing you’re missing.

**Comment by ryleu** [comment]
https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4

It's nearly time! Once this gets pushed out to stable, I would love to see AMD support here.

ETA: updates as of 20 hours ago suggest upstream support for OpenGL mesh shaders on RDNA2 or greater

**Comment by KarmaMarazu** [comment]
> ETA: updates as of 20 hours ago suggest upstream support for OpenGL mesh shaders on RDNA2 or greater

if u scroll up the 5th of june the same thing was said but yes RX 6000 will most likely have the support in their drivers

**Comment by FalsePattern** [comment]
https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4#issuecomment-2487628048

Looks like it's a direct implementation of the `GL_NV_mesh_shader` extension so it might even just work out of the box without needing any new bindings.

**Comment by MrPIRY** [comment]
https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4#issuecomment-2540423073 just have wait for the next driver update @MCRcortex I hope this won't be a pain

**Comment by MrPIRY** [comment]
https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4#issuecomment-2594436787 maybe next driver update or the update after that

**Comment by ItsLiyua** [comment]
Ii think the next driver update supports mesh shaders: https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/issues/4#issuecomment-2706311115

**Comment by ST-DDT** [comment]
The driver can already be downloaded from the AMD website. v25.3.1 (2025-03-06)
I dont have an AMD card yet, can somebody please try whether it works?

**Comment by MrPIRY** [comment]
> The driver can already be downloaded from the AMD website. v25.3.1 (2025-03-06) I dont have an AMD card yet, can somebody please try whether it works?

I have an amd GPU but we have to wait for cortex to add support for amd gpus that's if cortex actually wants to do it at all and if cortex hasn't used other opengl Nvidia exclusive tech

**Comment by ST-DDT** [comment]
AFAICT there is no Nvidia only check, only feature capability checks.

See https://github.com/MCRcortex/nvidium/blob/f2028b2ba9a5dc95d69e73a9eaeadcb447367d7d/src/main/java/me/cortex/nvidium/Nvidium.java#L29

**Comment by MrPIRY** [comment]
> AFAICT there is no Nvidia only check, only feature capability checks.
> 
> See
> 
> [nvidium/src/main/java/me/cortex/nvidium/Nvidium.java](https://github.com/MCRcortex/nvidium/blob/f2028b2ba9a5dc95d69e73a9eaeadcb447367d7d/src/main/java/me/cortex/nvidium/Nvidium.java#L29)
> 
> Line 29 in [f2028b2](/MCRcortex/nvidium/commit/f2028b2ba9a5dc95d69e73a9eaeadcb447367d7d)
> 
>  public static void checkSystemIsCapable() {

Yeah that's looking if the GPU is compatible using Nvidia mesh shaders, and I wasn't saying there was a Nvidia only check I was saying that cortex could have used more then just mesh shaders on Nvidias opengl version

**Comment by Poopooracoocoo** [comment]
Yes there are checks for many other NV extensions which AMD does not support

**Comment by MrPIRY** [comment]
> Yes there are checks for many other NV extensions which AMD does not support

will that be a problem if AMD doesn't have them? can Support for AMD still be added?

**Comment by Poopooracoocoo** [comment]
Well out of those, AMD only supports GL_ARB_sparse_buffer and GL_NV_mesh_shader. That's 2/6 extensions... A rewrite/new mod would be needed, wouldn't it? That's not exactly "adding support".

**Comment by MrPIRY** [comment]
> Well out of those, AMD only supports GL_ARB_sparse_buffer and GL_NV_mesh_shader. That's 2/6 extensions... A rewrite/new mod would be needed, wouldn't it? That's not exactly "adding support".

is there anyone that would actually make that a mod and btw what do the other five extensions do? are they that important?

**Comment by rafrafek** [comment]
Does anyone know if AMD currently implements or plans to implement all of the following required extensions?

- `GL_ARB_sparse_buffer`
- `GL_NV_bindless_multi_draw_indirect`
- `GL_NV_mesh_shader`
- `GL_NV_representative_fragment_test`
- `GL_NV_uniform_buffer_unified_memory`
- `GL_NV_vertex_buffer_unified_memory`

Are all of these extensions required? Are there any additional ones needed?

Also, are there any `GL_NV_*` extensions that could be replaced with ARB/KHR equivalents for vendor neutral support?

These are the extensions I found referenced in the source code:

```glsl
//#extension GL_NV_conservative_raster_underestimation : enable
#extension GL_ARB_shading_language_include : enable
#extension GL_KHR_shader_subgroup_arithmetic: require
#extension GL_KHR_shader_subgroup_ballot : require
#extension GL_KHR_shader_subgroup_basic : require
#extension GL_KHR_shader_subgroup_shuffle : require
#extension GL_KHR_shader_subgroup_vote : require
#extension GL_NV_bindless_texture : require
#extension GL_NV_fragment_shader_barycentric : require
#extension GL_NV_gpu_shader5 : require
#extension GL_NV_mesh_shader : require
#extension GL_NV_shader_buffer_load : require
```

**Comment by IGN-Styly** [comment]
> Does anyone know if AMD currently implements or plans to implement all of the following required extensions?
> 
> - `GL_ARB_sparse_buffer`
> - `GL_NV_bindless_multi_draw_indirect`
> - `GL_NV_mesh_shader`
> - `GL_NV_representative_fragment_test`
> - `GL_NV_uniform_buffer_unified_memory`
> - `GL_NV_vertex_buffer_unified_memory`
> 
> Are all of these extensions required? Are there any additional ones needed?
> 
> Also, are there any `GL_NV_*` extensions that could be replaced with ARB/KHR equivalents for vendor neutral support?
> 
> These are the extensions I found referenced in the source code:
> 
> ```glsl
> //#extension GL_NV_conservative_raster_underestimation : enable
> #extension GL_ARB_shading_language_include : enable
> #extension GL_KHR_shader_subgroup_arithmetic: require
> #extension GL_KHR_shader_subgroup_ballot : require
> #extension GL_KHR_shader_subgroup_basic : require
> #extension GL_KHR_shader_subgroup_shuffle : require
> #extension GL_KHR_shader_subgroup_vote : require
> #extension GL_NV_bindless_texture : require
> #extension GL_NV_fragment_shader_barycentric : require
> #extension GL_NV_gpu_shader5 : require
> #extension GL_NV_mesh_shader : require
> #extension GL_NV_shader_buffer_load : require
> ```
As far as I can tell from reading docs, a rewrite of some parts would be needed, if its even possible that is. Someone with nore opengl exp should check tho.
